packages/jest-runtime/src/index.ts

```diff
@@ -737,12 +740,70 @@ export default class Runtime {
     ) {
       return this.loadEsmModule(resolved, query);
     }
-
+    if (skipCjsModules) {
+      // @ts-expect-error - exiting
+      return;
+    }
     return this.loadCjsAsEsm(referencingIdentifier, resolved, context);
   }
 
+  private async checkForCyclicDependencies(
+    module: VMModule,
+    parentChain: Set<VMModule>,
+  ): Promise<boolean> {
+    if (
+      module.status !== 'unlinked' ||
+      this._hasCyclicDependencyMap.has(module)
+    ) {
+      return this._hasCyclicDependencyMap.get(module) as boolean;
+    }
+
+    // if the dependency resolves to a CJS module, it will be undefined
+    const dependencies: Array<VMModule | undefined> = await Promise.all(
+      (module.dependencySpecifiers || []).map(async (modSpec: VMModule) => {
+        const dependency = await this.resolveModule(
+          modSpec,
+          module.identifier,
+          module.context,
+          true,
+        );
+        return dependency;
+      }),
+    );
+
+    let hasCyclicDependency = false;
+    await Promise.all(
+      dependencies.filter(Boolean).map(async (possiblyCyclicMod: VMModule) => {
+        const isCyclic = parentChain.has(possiblyCyclicMod);
+        if (isCyclic) {
+          // if the found dependency is already in the parentChain
+          // this is a cyclic dependency. As such, everything between where
+          // that module is in the parentChain and the end of the chain
+          // (e.g., the current module) is also cyclic.
+          const chainInOrder = [...parentChain];
```

Comment:
is order in map guaranteed?

Reply:
I think more relevant is if `Set` order is guaranteed - in both cases (set and map) yes, they are.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set